src/PowerShellEditorServices.Engine/Services/TextDocument/CompletionResults.cs

```diff
@@ -0,0 +1,341 @@
+        {
+            Validate.IsNotNull("completionResult", completionResult);
+
+            // Some tooltips may have newlines or whitespace for unknown reasons
+            string toolTipText = completionResult.ToolTip;
+            if (toolTipText != null)
+            {
+                toolTipText = toolTipText.Trim();
+            }
+
+            return new CompletionDetails
+            {
+                CompletionText = completionResult.CompletionText,
+                ListItemText = completionResult.ListItemText,
+                ToolTipText = toolTipText,
+                SymbolTypeName = ExtractSymbolTypeNameFromToolTip(completionResult.ToolTip),
+                CompletionType =
+                    ConvertCompletionResultType(
+                        completionResult.ResultType)
+            };
+        }
+
+        internal static CompletionDetails Create(
+            string completionText,
+            CompletionType completionType,
+            string toolTipText = null,
+            string symbolTypeName = null,
+            string listItemText = null)
+        {
+            return new CompletionDetails
+            {
+                CompletionText = completionText,
+                CompletionType = completionType,
+                ListItemText = listItemText,
+                ToolTipText = toolTipText,
+                SymbolTypeName = symbolTypeName
+            };
+        }
+
+        #endregion
+
+        #region Public Methods
+
+        /// <summary>
+        /// Compares two CompletionResults instances for equality.
+        /// </summary>
+        /// <param name="obj">The potential CompletionResults instance to compare.</param>
+        /// <returns>True if the CompletionResults instances have the same details.</returns>
+        public override bool Equals(object obj)
+        {
+            CompletionDetails otherDetails = obj as CompletionDetails;
+            if (otherDetails == null)
+            {
+                return false;
+            }
+
+            return
+                string.Equals(this.CompletionText, otherDetails.CompletionText) &&
+                this.CompletionType == otherDetails.CompletionType &&
+                string.Equals(this.ToolTipText, otherDetails.ToolTipText) &&
+                string.Equals(this.SymbolTypeName, otherDetails.SymbolTypeName);
+        }
+
+        /// <summary>
+        /// Returns the hash code for this CompletionResults instance.
+        /// </summary>
+        /// <returns>The hash code for this CompletionResults instance.</returns>
+        public override int GetHashCode()
+        {
+            return
+                string.Format(
+                    "{0}{1}{2}{3}{4}",
+                    this.CompletionText,
+                    this.CompletionType,
+                    this.ListItemText,
+                    this.ToolTipText,
+                    this.SymbolTypeName).GetHashCode();
```

Comment:
What about case sensitivity?  Also ideally this wouldn't create an extra string.  Here's the template I usually go with:

```csharp
unchecked
{
    int hash = 17;
    hash = hash * 23 + CompletionText.GetHashCode();
    hash = hash * 23 + CompletionType.GetHashCode();
    hash = hash * 23 + ListItemText.GetHashCode();
    hash = hash * 23 + ToolTipText.GetHashCode();
    // If one can be null:
    hash = hash * 23 + SymbolTypeName?.GetHashCode() ?? 0;
    return hash;
}
```

(Taken from https://stackoverflow.com/a/263416)